File: benchmark/engine/agent.py
import re
import json
import asyncio
from typing import List, Dict, Any, Callable, Optional, Tuple
import os
import uuid

class BenchmarkAgent:
    def __init__(
        self,
        llm,
        model: str,
        workspace: str,
        tools: List[Any],
        log_path: Optional[str] = None,
        verbose: bool = True
    ):
        self.llm = llm
        self.model = model
        self.workspace = workspace
        self.tools = tools
        self.messages = []
        self.max_steps = 20
        self.log_path = log_path
        self.verbose = verbose
        self.tool_counts = {}
        self.total_tokens = 0

    def _log(self, message: str, end: str = "\n"):
        if self.log_path:
            with open(self.log_path, "a", encoding="utf-8") as f:
                f.write(message + end)

    async def run(self, task: str):
        self._log(f"=== NEW RUN for Model: {self.model} ===")
        self._log(f"Workspace: {self.workspace}")
        
        system_prompt = self._create_system_prompt()
        self._log("\n--- SYSTEM PROMPT ---\n" + system_prompt)
        self._log("\n--- USER TASK ---\n" + task)
        
        if self.verbose:
            print(f"Starting agent with task: {task}")
        
        self.messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": task}
        ]
        
        step = 0
        consecutive_retries = 0
        while step < self.max_steps:
            step += 1
            if self.verbose:
                print(f"\n--- Step {step} ---")
            self._log(f"\n--- STEP {step} ---")
            
            full_text = ""
            self._log("\n[LLM RESPONSE START]")
            usage_received = False
            async for event in self.llm.stream_chat(self.model, self.messages):
                if event["type"] == "token":
                    t = event["data"]
                    full_text += t
                    if self.verbose:
                        print(t, end="", flush=True)
                    self._log(t, end="")
                elif event["type"] == "usage":
                    self.total_tokens += event["data"].get("completion_tokens", 0)
                    usage_received = True
            
            if not usage_received and full_text:
                # Fallback: estimate tokens (approx 4 chars per token)
                estimated_tokens = max(1, len(full_text) // 4)
                self.total_tokens += estimated_tokens
                
            self._log("[LLM RESPONSE END]\n")
            
            if self.verbose:
                print("\n")
            
            tool_call = self._parse_tool_call(full_text)
            if tool_call:
                consecutive_retries = 0
                name, params = tool_call
                call_id = f"toolu-{uuid.uuid4().hex[:12]}"
                self.tool_counts[name] = self.tool_counts.get(name, 0) + 1
                if self.verbose:
                    print(f"Tool Call: {name}({params})")
                self._log(f"TOOL CALL DETECTED: {name}\nParameters: {json.dumps(params, indent=2)}")
                
                tool = next((t for t in self.tools if t.name() == name), None)
                if tool:
                    try:
                        result = await tool.execute(params, self.workspace)
                    except Exception as e:
                        result = f"Error: {str(e)}"
                else:
                    result = f"Error: Tool '{name}' not found"
                
                log_res = str(result)[:500] + "..." if len(str(result)) > 500 else str(result)
                self._log(f"TOOL RESULT:\n{log_res}")
                if self.verbose:
                    print(f"Tool Result: {str(result)[:200]}..." if len(str(result)) > 200 else f"Tool Result: {result}")
                
                # Format as <tool_response>
                content = result
                if isinstance(result, str):
                    try:
                        content = json.loads(result)
                    except:
                        content = {"message": result}
                
                response_data = {
                    "tool_call_id": call_id,
                    "name": name,
                    "content": content
                }
                
                self.messages.append({"role": "assistant", "content": full_text})
                self.messages.append({"role": "user", "content": f"<tool_response>\n{json.dumps(response_data)}\n</tool_response>"})
            elif "<tool_call>" in full_text:
                consecutive_retries += 1
                self.tool_counts["_malformed_"] = self.tool_counts.get("_malformed_", 0) + 1
                
                if consecutive_retries > 3:
                    self._log(f"MALFORMED TOOL CALL DETECTED (Retry limit reached: {consecutive_retries}/3)")
                    self._log("Too many malformed tool calls. stopping.")
                    return "Error: Too many malformed tool calls"
                
                self._log(f"MALFORMED TOOL CALL DETECTED (Retry {consecutive_retries}/3)")
                
                error_msg = "Error: Invalid tool call format. Ensure you provide a valid JSON object with 'name' and 'arguments' keys inside the <tool_call> tags."
                self.messages.append({"role": "assistant", "content": full_text})
                self.messages.append({"role": "user", "content": error_msg})
                continue
            else:
                if not full_text.strip():
                    consecutive_retries += 1
                    
                    if consecutive_retries <= 3:
                        self._log(f"EMPTY RESPONSE DETECTED (Retry {consecutive_retries}/3)")
                        if self.verbose:
                            print(f"Empty response, retrying ({consecutive_retries}/3)...")
                        # Skip adding to messages to keep it clean, just retry
                        continue
                    else:
                        self._log(f"EMPTY RESPONSE DETECTED (Retry limit reached: {consecutive_retries}/3)")
                        self._log("Too many empty responses. stopping.")
                        return "Error: Too many empty responses"

                self._log("Final Answer or no tool call received.")
                if self.verbose:
                    print("Final Answer received or no tool call.")
                return full_text
        
        if self.verbose:
            print("Max steps reached.")
        return "Error: Max steps reached"

    def _create_system_prompt(self) -> str:
        tools_desc = "\n".join([f"- {t.name()}: {t.description()}" for t in self.tools])
        return f"""You are an agentic coding assistant. Your workspace is: {self.workspace}

Available tools:
{tools_desc}

To call a tool, you MUST use the following XML format. DO NOT use your native tool calling syntax or special tokens like <|channel|> or <|message|>.
<tool_call>
{{"name": "tool_name", "arguments": {{"param1": "value1", "param2": "value2"}}}}
</tool_call>

Be precise and complete your tasks step by step. When you are done, provide a final summary of your work.
"""

    def _parse_tool_call(self, content: str) -> Optional[Tuple[str, Dict[str, Any]]]:
        # Try Strategy 1: XML Format (Default)
        match = re.search(r"<tool_call>(.*?)</(?:tool_call|tool_answer|tool_response)>", content, re.DOTALL)
        if match:
            try:
                inner = match.group(1).strip()
                data = json.loads(inner)
                name = data.get("name")
                params = data.get("arguments", {})
                if name:
                    return name, params
            except:
                pass

        # Try Strategy 2: Native Channel/Message Format (gpt-oss/llama-3 style)
        # Format: <|channel|>analysis to=container.exec code<|message|>{"cmd":["bash","-lc","ls -R"]}<|call|>
        if "<|message|>" in content:
            msg_match = re.search(r"<\|message\|>(.*?)(?:<\|call\|>|$)", content, re.DOTALL)
            if msg_match:
                try:
                    data = json.loads(msg_match.group(1).strip())
                    # If it's a bash command, map to run_command
                    if "cmd" in data and isinstance(data["cmd"], list):
                        # Simple mapping: join the list if it looks like a bash command
                        cmd_list = data["cmd"]
                        if len(cmd_list) > 1 and "bash" in cmd_list[0]:
                             # Extract actual command from -c or -lc
                             for i, arg in enumerate(cmd_list):
                                 if arg in ["-c", "-lc"] and i + 1 < len(cmd_list):
                                     return "run_command", {"command": cmd_list[i+1]}
                        # Fallback: join all
                        return "run_command", {"command": " ".join(cmd_list)}
                except:
                    pass

        # Try Strategy 3: Pure JSON block
        json_match = re.search(r"({.*})", content, re.DOTALL)
        if json_match:
            try:
                data = json.loads(json_match.group(1).strip())
                if "name" in data and "parameters" in data:
                    return data["name"], data["parameters"]
                if "tool" in data and "args" in data:
                    return data["tool"], data["args"]
                if "cmd" in data: # Direct command call
                    return "run_command", {"command": data["cmd"] if isinstance(data["cmd"], str) else " ".join(data["cmd"])}
            except:
                pass

        return None

File: vscode/src/core/agent.ts
import { Tool } from "./tools/base";
import { ToolCallParser } from "./parser";
import { PromptBuilder } from "./prompt";
import { v4 as uuidv4 } from "uuid";

export interface Message {
  role: "system" | "user" | "assistant";
  content: string;
  metadata?: any;
}

export interface StreamEvent {
  type: "token" | "usage" | "error" | "log" | "tool_started" | "tool_finished" | "final_answer" | "full_prompt";
  data?: any;
  message?: string;
  name?: string;
  parameters?: any;
  result?: any;
  call_id?: string;
}

export interface ModelPricing {
  prompt: number;     // USD per 1M tokens
  completion: number; // USD per 1M tokens
}

export interface ModelInfo {
  id: string;
  name?: string;
  pricing?: ModelPricing;
}

export interface LlmProvider {
  streamChat(model: string, messages: Message[]): AsyncIterable<StreamEvent>;
  fetchModels(): Promise<ModelInfo[]>;
  getPricing(model: string): ModelPricing | undefined;
}

export class Agent {
  private messages: Message[] = [];
  private stopped = false;
  private modifiedFiles: Set<string> = new Set();

  constructor(
    private llm: LlmProvider,
    private model: string,
    private workspaceName: string,
    private userName: string,
    private tools: Tool[],
    initialMessages: any[] = []
  ) {
    this.messages = initialMessages.map(m => ({
      role: m.role,
      content: typeof m.content === 'string' ? m.content : this._partsToString(m.content),
      metadata: m.metadata
    }));
  }

  private _partsToString(parts: any[]): string {
    return parts.map(p => p.content).join('\n');
  }

  async run(userPrompt: string, onEvent: (event: StreamEvent) => Promise<void>) {
    const systemPrompt = PromptBuilder.createSystemPrompt(
      this.tools.map(t => ({ name: t.name(), description: t.description(), schema: t.schema() })),
      this.workspaceName,
      this.userName
    );

    // Ensure system prompt is at the start
    if (this.messages.length === 0 || this.messages[0].role !== "system") {
      this.messages.unshift({ role: "system", content: systemPrompt });
    } else {
      // Update system prompt if it exists (in case workspace/tools changed, though unlikely here)
      this.messages[0].content = systemPrompt;
    }
    
    // Add user prompt if it's not already the last message
    const lastMessage = this.messages[this.messages.length - 1];
    if (!lastMessage || lastMessage.role !== "user" || lastMessage.content !== userPrompt) {
      this.messages.push({ role: "user", content: userPrompt });
    }

    await this.reasoningLoop(onEvent);
  }

  private async reasoningLoop(onEvent: (event: StreamEvent) => Promise<void>) {
    let totalSteps = 0;
    let emptyRetries = 0;
    let consecutiveNudges = 0;
    const maxEmptyRetries = 5;
    const maxConsecutiveNudges = 3;
    
    while (!this.stopped) {
      totalSteps++;
      const lastMsg = this.messages[this.messages.length - 1];
      await onEvent({ type: "log", message: `[Agent] Step ${totalSteps} | Last message: ${lastMsg.role} (${lastMsg.content.length} chars)` });
      
      // Log full history to log file via SessionManager (onEvent log type)
      const historySummary = this.messages.map(m => `[${m.role.toUpperCase()}] ${m.content.substring(0, 500)}${m.content.length > 500 ? '...' : ''}`).join('\n---\n');
      await onEvent({ type: "log", message: `[Agent] FULL CONTEXT SENT TO MODEL:\n${historySummary}` });
      await onEvent({ type: "full_prompt", data: this.messages });

      if (totalSteps > 50) {
        await onEvent({ type: "error", message: "Max steps reached" });
        break;
      }

      let fullText = "";
      let isSuppressing = false;
      let retries = 0;
      const maxRetries = 3;
      let success = false;

      while (retries < maxRetries && !success) {
        try {
          fullText = ""; // Reset for fresh attempt
          isSuppressing = false;
          
          for await (const event of this.llm.streamChat(this.model, this.messages)) {
            if (this.stopped) break;

            if (event.type === "token") {
              fullText += event.data;
              
              if (fullText.includes("<tool_call") || /<tool?\s*$/.test(fullText)) {
                if (!isSuppressing) {
                  await onEvent({ type: "log", message: "[Agent] Tool call detected, suppressing tokens from UI." });
                }
                isSuppressing = true;
              }

              if (!isSuppressing) {
                await onEvent(event);
              }
            } else if (event.type === "usage") {
              await onEvent(event);
            } else if (event.type === "log") {
              await onEvent(event);
            } else if (event.type === "error") {
              throw new Error(event.message || "Unknown provider error");
            }
          }

          // Check for very short or malformed response (streaming artifacts)
          const trimmed = fullText.trim();
          if (trimmed.length > 0 && (trimmed.length < 6 || trimmed === "<" || trimmed === "<<" || trimmed === "<<<")) {
             throw new Error(`Response too short or contains only artifacts: "${trimmed}"`);
          }

          success = true;
        } catch (e: any) {
          retries++;
          if (retries >= maxRetries) {
            await onEvent({ type: "error", message: `Failed after ${maxRetries} attempts: ${e.message}` });
            return;
          }
          // Wait 5 seconds before retry as requested
          await onEvent({ type: "log", message: `[Agent] Attempt ${retries} failed or was too short. Waiting 5s before retry ${retries + 1}/${maxRetries}... (${e.message})` });
          await new Promise(resolve => setTimeout(resolve, 5000));
        }
      }

      await onEvent({ type: "log", message: `[Agent] RAW LLM RESPONSE (Turn ${totalSteps}):\n${fullText}` });

      const { toolCall, error: parseError } = ToolCallParser.parse(fullText);
      
      // If tool_call tags are present but parsing failed, inform the LLM and retry
      if (!toolCall && (fullText.includes("<tool_call") || fullText.includes("<<<tool_call"))) {
        const displayError = parseError ? `Parsing error: ${parseError}` : "Unknown parsing error";
        await onEvent({ type: "log", message: `[Agent] Malformed tool call detected. ${displayError}` });
        
        // Log full malformed response for debugging
        await onEvent({ type: "log", message: `[Agent] FULL MALFORMED RESPONSE:\n${fullText}` });

        // Truncate huge malformed responses to avoid context bloat
        let historyContent = fullText;
        if (fullText.length > 5000) {
          const thoughtRegex = /<thought>([\s\S]*?)<\/thought>/;
          const thoughtMatch = fullText.match(thoughtRegex);
          const thought = thoughtMatch ? thoughtMatch[0] : "";
          historyContent = `${thought}\n\n<tool_call>\n... [TRUNCATED DUE TO SIZE: ${fullText.length} chars] ...\n</tool_call>`;
        }

        this.messages.push({ role: "assistant", content: historyContent });
        
        const nudge = `❌ Malformed tool call detected. ${displayError}.
Please ensure your tool call is a valid JSON object wrapped in <tool_call> tags.
- Use DOUBLE QUOTES for all keys and string values.
- Escape all backslashes and double quotes inside string values (e.g. \\" and \\\\).
- Do not include trailing commas.
- If you are writing a large file, double check the JSON escaping or use 'run_command' with 'cat' and a HEREDOC if appropriate.

Example: <tool_call>{"name": "tool_name", "arguments": {"param": "value"}}</tool_call>`;

        this.messages.push({ 
          role: "user", 
          content: nudge
        });

        // Emit events so the UI knows about the failure and the nudge
        await onEvent({ type: "log", message: `[Agent] Added malformed assistant response to history.` });
        await onEvent({ type: "log", message: `[Agent] Internal User Nudge: ${nudge}` });
        
        continue;
      }

      if (toolCall) {
        const callId = `toolu-${uuidv4().substring(0, 12)}`;
        await onEvent({
          type: "tool_started",
          name: toolCall.name,
          parameters: toolCall.arguments,
          call_id: callId
        });
        await onEvent({ type: "log", message: `Executing tool: ${toolCall.name} with arguments: ${JSON.stringify(toolCall.arguments)}` });

        const tool = this.tools.find(t => t.name() === toolCall.name);
        let result: any;
        const toolStartTime = Date.now();
        if (tool) {
          try {
            result = await tool.execute(toolCall.arguments);
            if (['write_file', 'edit_file'].includes(toolCall.name) && toolCall.arguments.path) {
              this.modifiedFiles.add(toolCall.arguments.path);
            }
          } catch (e: any) {
            result = `Error: ${e.message}`;
          }
        } else {
          result = `Error: Tool '${toolCall.name}' not found`;
        }
        const toolDuration = Date.now() - toolStartTime;

        await onEvent({
          type: "tool_finished",
          name: toolCall.name,
          result: result,
          call_id: callId,
          parameters: { duration: toolDuration } // Pass duration back
        });
        await onEvent({ type: "log", message: `Tool ${toolCall.name} finished. Result length: ${typeof result === 'string' ? result.length : 'N/A'}` });

        // Inject the generated callId into the tool_call tag so the history is consistent
        const callIdTag = ` id="${callId}"`;
        const updatedFullText = fullText.replace('<tool_call>', `<tool_call${callIdTag}>`);
        
        this.messages.push({ role: "assistant", content: updatedFullText });
        this.messages.push({
          role: "user",
          content: PromptBuilder.formatToolResult(toolCall.name, result, callId)
        });
        consecutiveNudges = 0; // Reset on tool call
      } else {
        // Strip all thinking/thought content to check what's left
        const thoughtRegex = /<(?:thought|thinking)>[\s\S]*?<\/(?:thought|thinking)>/g;
        const textWithoutThoughts = fullText.replace(thoughtRegex, '').trim();
        
        // Case 1: Completely empty response — silent retry without polluting history
        if (fullText.trim().length === 0) {
          emptyRetries++;
          if (emptyRetries >= maxEmptyRetries) {
            await onEvent({ type: "error", message: `LLM returned empty responses ${maxEmptyRetries} times. Stopping.` });
            break;
          }
          await onEvent({ type: "log", message: `[Agent] Empty response from LLM. Retrying silently (${emptyRetries}/${maxEmptyRetries})...` });
          await new Promise(resolve => setTimeout(resolve, 1000));
          continue;
        }

        // Case 2: Only thoughts, no action or text — retry without adding nudge to history
        if (textWithoutThoughts.length === 0) {
          emptyRetries++;
          if (emptyRetries >= maxEmptyRetries) {
            await onEvent({ type: "error", message: `LLM returned only thoughts ${maxEmptyRetries} times. Stopping.` });
            break;
          }
          await onEvent({ type: "log", message: `[Agent] LLM only provided thoughts without action. Retrying silently (${emptyRetries}/${maxEmptyRetries})...` });
          // Don't push to history — just retry as if nothing happened
          await new Promise(resolve => setTimeout(resolve, 500));
          continue;
        }

        // Case 3: Task explicitly finished
        if (fullText.includes('<task_finished')) {
          await onEvent({ type: "log", message: "[Agent] Task finished tag detected. Stopping loop." });
          await onEvent({ 
            type: "final_answer", 
            data: fullText,
            parameters: { modifiedFiles: Array.from(this.modifiedFiles) }
          });
          break;
        }

        // Case 4: LLM responded with actual content but no <task_finished />
        // If we haven't nudged too many times, try to encourage it to continue.
        if (consecutiveNudges < maxConsecutiveNudges) {
          consecutiveNudges++;
          await onEvent({ type: "log", message: `[Agent] LLM responded with content but no <task_finished />. Nudging (${consecutiveNudges}/${maxConsecutiveNudges})...` });
          
          this.messages.push({ role: "assistant", content: fullText });
          const nudge = "You provided a response but did not include the <task_finished /> tag and did not call any tool. If you are finished, please provide a final summary and include <task_finished />. If you have more work to do, please proceed with the next step or tool call.";
          this.messages.push({ role: "user", content: nudge });
          
          continue;
        }

        // Only treat as final answer after exceeding max nudges
        await onEvent({ type: "log", message: `[Agent] Exceeded max nudges without task completion tag. Treating as final answer.` });
        await onEvent({ 
          type: "final_answer", 
          data: fullText,
          parameters: { modifiedFiles: Array.from(this.modifiedFiles) }
        });
        break;
      }
    }
  }

  stop() {
    this.stopped = true;
  }
}
